pkg/providers/vsphere/vmprovider_vm_snapshot.go

```diff
@@ -819,3 +824,35 @@ func (vs *vSphereVMProvider) reconcileCurrentSnapshot(
 
 	return nil
 }
+
+func (vs *vSphereVMProvider) reconcilePostRevertPowerState(
+	vmCtx pkgctx.VirtualMachineContext,
+	snapObj *vimtypes.ManagedObjectReference) error {
+
+	snapshot := vmlifecycle.FindSnapshotInTree(vmCtx.MoVM.Snapshot.RootSnapshotList, snapObj.Value)
+	if snapshot == nil {
+		// weird situation here
+		// the snapshot being processed doesn't exist in the snapshot tree
+		// this shouldn't be happening
+		return fmt.Errorf("snapshot %s not found in VM Snapshot Tree", snapObj.Value)
+	}
+
+	// vpxd sets the power state of the snapshot to Off if a VM which is On
+	// is snapshotted without the memory state. We don't need to separately
+	// handle that case here.
+	switch snapshot.State {
+	case vimtypes.VirtualMachinePowerStatePoweredOn:
+		vmCtx.VM.Spec.PowerState = vmopv1.VirtualMachinePowerStateOn
+	case vimtypes.VirtualMachinePowerStatePoweredOff:
+		vmCtx.VM.Spec.PowerState = vmopv1.VirtualMachinePowerStateOff
+	case vimtypes.VirtualMachinePowerStateSuspended:
+		vmCtx.VM.Spec.PowerState = vmopv1.VirtualMachinePowerStateSuspended
+	default:
+		// Unknown state, log and default to off
```

Comment:
You are not just logging, but returning an error.  I agree that defaulting to powered off makes sense.